pkg/storage/backend-storage/backend-storage.go

```diff
@@ -71,24 +76,115 @@ func IsBackendStorageNeededForVM(vm *corev1.VirtualMachine) bool {
 	return HasPersistentTPMDevice(&vm.Spec.Template.Spec)
 }
 
-func CreateIfNeeded(vmi *corev1.VirtualMachineInstance, clusterConfig *virtconfig.ClusterConfig, client kubecli.KubevirtClient) error {
+type BackendStorage struct {
+	client          kubecli.KubevirtClient
+	clusterConfig   *virtconfig.ClusterConfig
+	scStore         cache.Store
+	accessModeCache map[string]v1.PersistentVolumeAccessMode
+	cacheLock       sync.Mutex
+}
+
+func NewBackendStorage(client kubecli.KubevirtClient, clusterConfig *virtconfig.ClusterConfig, scStore cache.Store) *BackendStorage {
+	return &BackendStorage{
+		client:          client,
+		clusterConfig:   clusterConfig,
+		scStore:         scStore,
+		accessModeCache: map[string]v1.PersistentVolumeAccessMode{},
+	}
+}
+
+func (bs *BackendStorage) getStorageClass() (string, error) {
+	storageClass := bs.clusterConfig.GetVMStateStorageClass()
+	if storageClass != "" {
+		return storageClass, nil
+	}
+
+	for _, obj := range bs.scStore.List() {
+		sc := obj.(*storagev1.StorageClass)
+		if sc.Annotations["storageclass.kubernetes.io/is-default-class"] == "true" {
+			return sc.Name, nil
+		}
+	}
+
+	return "", fmt.Errorf("no default storage class found")
+}
+
+func (bs *BackendStorage) getMode(storageClass string, mode v1.PersistentVolumeMode) v1.PersistentVolumeAccessMode {
+	var storageProfile *v1beta1.StorageProfile
+	bs.cacheLock.Lock()
+	defer bs.cacheLock.Unlock()
+	accessMode, exists := bs.accessModeCache[storageClass]
+	if exists {
+		return accessMode
+	}
+
+	// status.storageClass is unfortunately not a supported field label.
+	// If storage profiles are guaranteed to have the same name as their storage class, this could just be a get.
+	sps, err := bs.client.CdiClient().CdiV1beta1().StorageProfiles().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		// CDI not present most likely
+		log.Log.Reason(err).Info("couldn't list storage profiles, defaulting to RWX")
+		return v1.ReadWriteMany
+	}
+	for _, sp := range sps.Items {
+		if sp.Status.StorageClass != nil && *sp.Status.StorageClass == storageClass {
+			storageProfile = &sp
+			break
+		}
+	}
+	if storageProfile == nil {
+		log.Log.Infof("no storage profile found for %s, defaulting to RWX", storageClass)
+		return v1.ReadWriteMany
+	}
+
+	bs.accessModeCache[storageClass] = v1.ReadWriteMany
+	if storageProfile.Status.ClaimPropertySets == nil || len(storageProfile.Status.ClaimPropertySets) == 0 {
+		log.Log.Infof("no ClaimPropertySets in storage profile %s, defaulting to RWX", storageProfile.Name)
+		return v1.ReadWriteMany
+	}
+
+	foundrwo := false
+	for _, property := range storageProfile.Status.ClaimPropertySets {
+		if property.VolumeMode != nil && *property.VolumeMode == mode {
```

Comment:
nit: can we reverse the condition and put a continue. This could avoid too many nested levels